plugins/bnr-dp/fu-bnr-dp-firmware.c

```diff
@@ -0,0 +1,595 @@
+				    "invalid or missing firmware header element");
+	}
+
+	if (!fu_bnr_dp_firmware_attribute_parse_u64(root, "Dev", &self->device_id, error))
+		return FALSE;
+
+	if (!fu_bnr_dp_firmware_attribute_parse_u64(root, "Ver", &tmp_u64, error))
+		return FALSE;
+	fu_firmware_set_version_raw(FU_FIRMWARE(self), tmp_u64);
+
+	self->usage = fu_bnr_dp_firmware_attribute_parse_string(root, "Use", error);
+	if (self->usage == NULL)
+		return FALSE;
+	if (g_strcmp0(self->usage, "fw") != 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "unsupported usage string in XML header: '%s'",
+			    self->usage);
+		return FALSE;
+	}
+
+	tmp_str = fu_bnr_dp_firmware_attribute_parse_string(root, "Fct", error);
+	if (tmp_str == NULL || strlen(tmp_str) != 1)
+		return FALSE;
+	self->function = tmp_str[0];
+	/* function compatibility check */
+	if (self->function != '_') {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "unexpected function (Fct) value in XML header: '%c' (0x%hX)",
+			    self->function,
+			    self->function);
+		return FALSE;
+	}
+
+	if (!fu_bnr_dp_firmware_attribute_parse_u64(root, "Var", &self->variant, error))
+		return FALSE;
+
+	if (!fu_bnr_dp_firmware_attribute_parse_u64(root, "Len", &self->payload_length, error))
+		return FALSE;
+
+	g_free(tmp_str);
+	tmp_str = fu_bnr_dp_firmware_attribute_parse_string(root, "Chk", error);
+	if (tmp_str == NULL ||
+	    !(g_str_has_prefix(tmp_str, "0x") || g_str_has_prefix(tmp_str, "0X")))
+		return FALSE;
+	if (!g_ascii_string_to_unsigned(tmp_str + 2, 16, 0, G_MAXUINT16, &tmp_u64, error))
+		return FALSE;
+	self->payload_checksum = (guint16)tmp_u64;
+
+	self->material = fu_bnr_dp_firmware_attribute_parse_string(root, "Mat", error);
+	if (self->material == NULL)
+		return FALSE;
+
+	fw_creation_date = fu_bnr_dp_firmware_attribute_parse_string(root, "Date", error);
+	if (fw_creation_date != NULL)
+		g_info("firmware creation date (dd.mm.yyyy): %s", fw_creation_date);
+
+	fw_comment = fu_bnr_dp_firmware_attribute_parse_string(root, "Rem", error);
+	if (fw_comment != NULL && strlen(fw_comment) > 0)
+		g_info("firmware comment: %s", fw_comment);
+
+	return TRUE;
+}
+
+static guint16
+fu_bnr_dp_firmware_checksum_finish(guint16 csum)
+{
+	return ~csum + 1;
+}
+
+static gboolean
+fu_bnr_dp_firmware_stream_checksum(GInputStream *stream, guint16 *csum, GError **error)
+{
+	if (!fu_input_stream_compute_sum16(stream, csum, error))
+		return FALSE;
+	*csum = fu_bnr_dp_firmware_checksum_finish(*csum);
+
+	return TRUE;
+}
+
+static guint16
+fu_bnr_dp_firmware_buf_checksum(const guint8 *buf, gsize bufsz)
+{
+	guint16 csum;
+
+	csum = fu_sum16(buf, bufsz);
+	csum = fu_bnr_dp_firmware_checksum_finish(csum);
+
+	return csum;
+}
+
+static gboolean
+fu_bnr_dp_firmware_payload_parse(FuBnrDpFirmware *self,
+				 GInputStream *stream,
+				 gsize payload_offset,
+				 GError **error)
+{
+	g_autoptr(GInputStream) payload_stream = NULL;
```

Comment:
just by convention, the g_autoptr() stuff normally comes at the end of the initialization block.